### PR TITLE
feat: snp backend sends CRL for attestation

### PR DESCRIPTION
**Pull request by rjzak**
Refs:
* SNP companion to https://github.com/enarx/enarx/pull/2393
* https://github.com/enarx/enarx/issues/2389

Signed-off-by: Richard Zak <richard@profian.com>

### CODE REVIEW

**Review by npmccallum — COMMENTED** [review]
I'm confused by the different versions of the `SnpEvidence` struct.

`src/backend/sev/snp/vcek.rs`:

```rust
#[derive(Sequence)]
pub struct SnpEvidence {
    pub vcek: Document,
    pub crl: Document,
}
```

`tests/crates/enarx_exec_tests/src/bin/sev_attestation.rs`:

```rust
#[derive(Sequence)]
struct SnpEvidence<'a> {
    vcek: Document,
    crl: Vec<CertificateList<'a>>,
}
```

Then, in Steward we have:

```
#[derive(Clone, Debug, PartialEq, Eq, Sequence)]
pub struct Evidence<'a> {
    pub vcek: Certificate<'a>,

    #[asn1(type = "OCTET STRING")]
    pub report: &'a [u8],
}
```

These are three subtly different versions of the same idea. Is that really what we want?

**Comment by rjzak** [comment]
@npmccallum 

1. Because this code is reading the CRL from disk, so having it as a `der::Document` allows for the Der encoder to add the bytes as an existing ASN1 structure rather than re-encoding an ASN1 (resulting in double encoding). This also validates that the ASN1 encoding is valid. It would be unnecessary to decode the DER to a Vec of CRLs, just to re-encode it again.
2. The unit test validates the CRL by actually decoding it into a Vector of CRLs.
3. Steward is a bit different from that:

```rust
#[derive(Clone, Debug, PartialEq, Eq, Sequence)]
pub struct Certificates<'a> {
    pub vcek: Certificate<'a>,
    pub crl: Vec<CertificateList<'a>>,
}

#[derive(Clone, Debug, PartialEq, Eq, Sequence)]
pub struct Evidence<'a> {
    pub crts: Certificates<'a>,

    #[asn1(type = "OCTET STRING")]
    pub report: &'a [u8],
}
```
The `Certificates` struct holds both the CRL and the vcek.

**Comment by npmccallum** [comment]
Stepping back to think about the flow here:

1. The host needs to provide the VCEK and the CRLs to the shim.
2. The shim needs to combine the VCEK and CRLs with the attestation report for return to the exec layer.
3. HOWEVER, we don't want to teach the shims how to do ASN.1 parsing.

This currently works (without CRLs) by using the GetSnpVcek sallyport call to the host to get the VCEK. Then it combines the VCEK and report by hand-crafting some ASN.1 bytes.

I think @haraldh needs to look at this.

**Comment by rjzak** [comment]
I don't believe the shims are involved here, this is the backend on the host.

**Review by haraldh — CHANGES_REQUESTED** [review]
Suggested changes:
https://github.com/haraldh/enarx/commit/8d7cbd8326a7c4566d9e302247deabcc914680b6

File: src/backend/sev/mod.rs [hunk 1]
```diff
@@ -66,8 +67,10 @@ impl KeepPersonality for SnpKeepPersonality {
                         .map_err(io::Error::from_raw_os_error)
                         .context("snp::enarxcall deref")?
                 };
-                let mut vcek_reader = get_vcek_reader()?;
-                *ret = std::io::copy(&mut vcek_reader, &mut vcek_buf)? as _;
+                //let mut vcek_reader = get_vcek_reader()?;
```
[the remark this answers is not in the document]
Comment (rjzak) [reply src/backend/sev/mod.rs]: Removed

**Review by haraldh — APPROVED** [review]

**Pull request merged**